samples/nrf9160/cloud_simple/src/main.c

```diff
@@ -0,0 +1,216 @@
+#include <lte_lc.h>
+#include <net/cloud.h>
+#include <net/socket.h>
+#include <dk_buttons_and_leds.h>
+
+static struct cloud_backend *cloud_backend;
+static struct k_delayed_work cloud_update_work;
+
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib recoverable error: %u\n", err);
+}
+
+static void cloud_update_work_fn(struct k_work *work)
+{
+	int err;
+
+	printk("Publishing message: %s\n", CONFIG_CLOUD_MESSAGE);
+
+	struct cloud_msg msg = {
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_TOPIC_MSG,
+		.buf = CONFIG_CLOUD_MESSAGE,
+		.len = sizeof(CONFIG_CLOUD_MESSAGE)
+	};
+
+	err = cloud_send(cloud_backend, &msg);
+	if (err) {
+		printk("cloud_send failed, error: %d\n", err);
+	}
+
+#if defined(CONFIG_CLOUD_PUBLICATION_SEQUENTIAL)
+	k_delayed_work_submit(
+			&cloud_update_work,
+			K_SECONDS(CONFIG_CLOUD_MESSAGE_PUBLICATION_INTERVAL));
+#endif
+}
+
+void cloud_event_handler(const struct cloud_backend *const backend,
+			 const struct cloud_event *const evt,
+			 void *user_data)
+{
+	ARG_UNUSED(user_data);
+	ARG_UNUSED(backend);
+
+	switch (evt->type) {
+	case CLOUD_EVT_CONNECTED:
+		printk("CLOUD_EVT_CONNECTED\n");
+		break;
+	case CLOUD_EVT_READY:
+		printk("CLOUD_EVT_READY\n");
+#if defined(CONFIG_CLOUD_PUBLICATION_SEQUENTIAL)
+		k_delayed_work_submit(&cloud_update_work, K_NO_WAIT);
+#endif
+		break;
+	case CLOUD_EVT_DISCONNECTED:
+		printk("CLOUD_EVT_DISCONNECTED\n");
+		break;
+	case CLOUD_EVT_ERROR:
+		printk("CLOUD_EVT_ERROR\n");
+		break;
+	case CLOUD_EVT_DATA_SENT:
+		printk("CLOUD_EVT_DATA_SENT\n");
+		break;
+	case CLOUD_EVT_DATA_RECEIVED:
+		printk("CLOUD_EVT_DATA_RECEIVED\n");
+		printk("Data received from cloud: %s\n", evt->data.msg.buf);
+		break;
+	case CLOUD_EVT_PAIR_REQUEST:
+		printk("CLOUD_EVT_PAIR_REQUEST\n");
+		break;
+	case CLOUD_EVT_PAIR_DONE:
+		printk("CLOUD_EVT_PAIR_DONE\n");
+		break;
+	case CLOUD_EVT_FOTA_DONE:
+		printk("CLOUD_EVT_FOTA_DONE\n");
+		break;
+	default:
+		printk("Unknown cloud event type: %d\n", evt->type);
+		break;
+	}
+}
+
+static void work_init(void)
+{
+	k_delayed_work_init(&cloud_update_work, cloud_update_work_fn);
+}
+
+static void modem_configure(void)
+{
+#if defined(CONFIG_BSD_LIBRARY)
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already turned on
+		 * and connected.
+		 */
+	} else {
+		int err;
+
+		printk("Connecting to LTE network. ");
+		printk("This may take several minutes.\n");
+
+		err = lte_lc_init_and_connect();
+		if (err) {
+			printk("LTE link could not be established.\n");
+		}
+
+		printk("Connected to LTE network\n");
+
+		err = lte_lc_psm_req(true);
+		if (err) {
+			printk("lte_lc_psm_req, error: %d\n", err);
+		}
+
+		printk("PSM mode requested\n");
+	}
+#endif
+}
+
+#if defined(CONFIG_CLOUD_PUBLICATION_BUTTON_PRESS)
+static void button_handler(u32_t button_states, u32_t has_changed)
+{
+	if ((has_changed & DK_BTN1_MSK)) {
+		k_delayed_work_submit(&cloud_update_work, K_SECONDS(2));
```

Comment:
Any reason for waiting 2 seconds? It could ble scheduled with no delay?

Reply:
The button handler is sensitive to any changes in button state meaning that a push and release will schedule 2 items. The 2 second duration is to prevent two delayed work items being added to workqueue. If the button is released within two seconds the initial workitem is canceled and only one item is scheduled.